packages/libsql/src/lib/Category.ts

```diff
@@ -0,0 +1,417 @@
+    })
+    return [Token.create(version), state]
+  }
+
+  async loadLast<Event, State>(
+    streamName: string,
+    decode: Decode<Event>,
+    fold: (state: State, events: Event[]) => State,
+    initial: State,
+  ): Promise<[StreamToken, State]> {
+    const [version, events] = await Read.loadLastEvent(this.conn.read, streamName)
+    trace.getActiveSpan()?.setAttributes({
+      [Tags.loaded_count]: 1,
+      [Tags.read_version]: Number(version),
+    })
+    return [Token.create(version), fold(initial, keepMap(events, decode))]
+  }
+
+  async loadState<Event, State>(
+    streamName: string,
+    decode: Decode<Event>,
+    fold: (state: State, events: Event[]) => State,
+    initial: State,
+  ): Promise<[StreamToken, State]> {
+    const snapshot = await this.conn.read.readSnapshot(streamName)
+    // An end-user might add RollingState to a category after the fact, we should handle this gracefully
+    if (snapshot == null) return this.loadBatched(streamName, decode, fold, initial)
+    const [etag, event] = snapshot
+    const decoded = decode(event)
+    // An end-user might add RollingState to a category after the fact, we should handle this gracefully
+    if (!decoded) return this.loadBatched(streamName, decode, fold, initial)
+    const version = BigInt(event.index)
+    const state = fold(initial, [decoded])
+    const span = trace.getActiveSpan()
+    span?.setAttributes({
+      [Tags.loaded_count]: 1,
+      [Tags.snapshot_version]: Number(version),
+    })
+    return [Token.create(version, etag), state]
+  }
+
+  async loadSnapshot<Event, State>(
+    streamName: string,
+    decode: Decode<Event>,
+    fold: (state: State, events: Event[]) => State,
+    initial: State,
+    isOrigin: (e: Event) => boolean,
+  ): Promise<[StreamToken, State]> {
+    const snapshot = await this.conn.read.readSnapshot(streamName)
+    // An end-user might add snapshotting to a category after the fact, in which case there might not be a snapshot
+    // in all other cases the snapshot is guaranteed to exist if there are any events in the stream
+    if (snapshot == null) return this.loadBatched(streamName, decode, fold, initial)
+    const [_, event] = snapshot
+    const decoded = decode(event)
+    // The snapshot type may have changed, in which case we should ignore it and load a fresh state
+    if (!decoded || !isOrigin(decoded)) return this.loadBatched(streamName, decode, fold, initial)
+    const version = event.index
+    const state = fold(initial, [decoded])
+    const span = trace.getActiveSpan()
+    span?.setAttributes({
+      [Tags.loaded_count]: 1,
+      [Tags.snapshot_version]: Number(version),
+    })
+    return [Token.create(version), state]
+  }
+
+  async reload<Event, State>(
+    streamName: Equinox.StreamName,
+    token: StreamToken,
+    decode: Decode<Event>,
+    fold: (state: State, events: Event[]) => State,
+    state: State,
+  ): Promise<[StreamToken, State]> {
+    let streamVersion = Token.version(token)
```

Comment:
stateVersion?

Reply:
side-note if the cached state is behind and the stream has snapshots, then doing a single read of that would be a better first step than loading potentially lots of events

I assume latestevent mode already does something itself to avoid this too (it cant in esdb/sss and I'm flaggign in case that logic is being carried forward suboptimally)

Reply:
That was definitely forwarded onwards to mdb as well, will do a comprehensive optimisiation after merging this